diff --git a/operators/tokenizer/basic_tokenizer.hpp b/operators/tokenizer/basic_tokenizer.hpp
index 6ea80a3..e65b2ee 100644
--- a/operators/tokenizer/basic_tokenizer.hpp
+++ b/operators/tokenizer/basic_tokenizer.hpp
@@ -3,8 +3,8 @@
 
 #pragma once
 
-#include "ocos.h"
-#include "string_utils.h"
+#include <vector>
+#include <string_view>
 #include "ustring.h"
 
 class BasicTokenizer {
@@ -20,23 +20,3 @@ class BasicTokenizer {
   bool tokenize_punctuation_;
   bool remove_control_chars_;
 };
-
-struct KernelBasicTokenizer{
-  template <typename T>
-  KernelBasicTokenizer(const T& dict) {
-    bool do_lower_case = dict.TryToGetAttributeWithDefault("do_lower_case", true);
-    bool tokenize_chinese_chars = dict.TryToGetAttributeWithDefault("tokenize_chinese_chars", true);
-    bool strip_accents = dict.TryToGetAttributeWithDefault("strip_accents", false);
-    bool tokenize_punctuation = dict.TryToGetAttributeWithDefault("tokenize_punctuation", false);
-    bool remove_control_chars = dict.TryToGetAttributeWithDefault("remove_control_chars", true);
-
-    tokenizer_ = std::make_shared<BasicTokenizer>(do_lower_case, tokenize_chinese_chars, strip_accents,
-                                                  tokenize_punctuation, remove_control_chars);
-  }
-
-  void Compute(std::string_view input,
-               ortc::Tensor<std::string>& output) const;
-
- private:
-  std::shared_ptr<BasicTokenizer> tokenizer_;
-};
diff --git a/operators/tokenizer/bert_tokenizer.hpp b/operators/tokenizer/bert_tokenizer.hpp
index 8be74c9..c35519c 100644
--- a/operators/tokenizer/bert_tokenizer.hpp
+++ b/operators/tokenizer/bert_tokenizer.hpp
@@ -3,14 +3,14 @@
 
 #pragma once
 
-#include "ocos.h"
-#include "ustring.h"
-#include "string_utils.h"
-#include "string_tensor.h"
-#include "basic_tokenizer.hpp"
-
+#include <memory>
+#include <string>
+#include <string_view>
 #include <unordered_map>
+#include <vector>
 #include <list>
+#include "ustring.h"
+#include "basic_tokenizer.hpp"
 
 class BertTokenizerVocab final {
  public:
@@ -94,26 +94,3 @@ class BertTokenizer final {
   std::unique_ptr<BasicTokenizer> basic_tokenizer_;
   std::unique_ptr<WordpieceTokenizer> wordpiece_tokenizer_;
 };
-
-struct KernelBertTokenizer : BaseKernel {
-  KernelBertTokenizer(const OrtApi& api, const OrtKernelInfo& info);
-  void Compute(const ortc::Tensor<std::string>& input,
-               ortc::Tensor<int64_t>& output,   // input_ids
-               ortc::Tensor<int64_t>& output1,  // token_type_ids
-               ortc::Tensor<int64_t>& output2,  // attention_mask
-               std::optional<ortc::Tensor<int64_t>*> offset_mapping) const;
-  using OffsetMappingType = std::list<std::pair<size_t, size_t>>;
-
- protected:
-  std::unique_ptr<BertTokenizer> tokenizer_;
-};
-
-struct KernelHfBertTokenizer : KernelBertTokenizer {
-  KernelHfBertTokenizer(const OrtApi& api, const OrtKernelInfo& info);
-  using OffsetMappingType = std::list<std::pair<size_t, size_t>>;
-  void Compute(const ortc::Tensor<std::string>& input,
-               ortc::Tensor<int64_t>& output,                  // input_ids
-               ortc::Tensor<int64_t>& output1,                 // attention_mask
-               std::optional<ortc::Tensor<int64_t>*> output2,  // token_type_ids
-               std::optional<ortc::Tensor<int64_t>*> offset_mapping) const;
-};
