--- src/braft/replicator.cpp
+++ src/braft/replicator.cpp
@@ -428,7 +428,8 @@ static void* _on_rpc_returned(void* arg) {
             r->_notify_on_caught_up(EPERM, true);
             butil::Status status;
             status.set_error(EHIGHERTERMRESPONSE, "Leader receives higher term "
-                    "%s from peer:%s", response->GetTypeName().c_str(), r->_options.peer_id.to_string().c_str());
+                    "%.*s from peer:%s", static_cast<int>(response->GetTypeName().size()),
+                    response->GetTypeName().data(), r->_options.peer_id.to_string().c_str());
             r->_destroy();
             node_impl->increase_term_to(response->term(), status);
             node_impl->Release();
